File: matterjs-server-main/packages/matter-server/test/helpers/ProcessHelpers.ts
/**
 * @license
 * Copyright 2025-2026 Open Home Foundation
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Helper functions for process and server management in tests.
 */

import { ChildProcess, spawn } from "child_process";
import { mkdir, rm } from "fs/promises";
import nodeProcess from "node:process";
import { tmpdir } from "os";
import { join } from "path";
import WebSocket from "ws";

// Constants - use non-default ports to avoid conflicts with any real Matter
// server/device that may already be bound to the production defaults on the
// host (ws server 5580, Matter protocol 5540).
export const SERVER_PORT = 5590;
export const SERVER_WS_URL = `ws://localhost:${SERVER_PORT}/ws`;
export const DEVICE_PORT = 5550;
export const MANUAL_PAIRING_CODE = "34970112332";
export const DEVICE_PASSCODE = 20202021;
export const DEVICE_DISCRIMINATOR = 3840;

/**
 * Creates temporary storage directories for server and device, plus a log file path.
 * Returns paths that should be cleaned up after tests.
 */
export async function createTempStoragePaths(): Promise<{
    serverStoragePath: string;
    deviceStoragePath: string;
    logFilePath: string;
}> {
    const timestamp = Date.now();
    const serverStoragePath = join(tmpdir(), `matter-test-server-${timestamp}`);
    const deviceStoragePath = join(tmpdir(), `matter-test-device-${timestamp}`);

    await mkdir(serverStoragePath, { recursive: true });
    await mkdir(deviceStoragePath, { recursive: true });

    // Log file lives inside server storage dir so it is cleaned up automatically.
    const logFilePath = join(serverStoragePath, "server.log");

    return { serverStoragePath, deviceStoragePath, logFilePath };
}

/**
 * Cleans up temporary storage directories.
 */
export async function cleanupTempStorage(serverStoragePath: string, deviceStoragePath: string): Promise<void> {
    try {
        await rm(serverStoragePath, { recursive: true, force: true });
        await rm(deviceStoragePath, { recursive: true, force: true });
    } catch {
        // Ignore cleanup errors
    }
}

/**
 * Starts the Matter server process, optionally writing logs to a file.
 */
export function startServer(
    storagePath: string,
    logFilePath?: string,
    logLevel = process.env.MATTER_LOG_LEVEL ?? "info",
    enableTestNetDcl = true,
): ChildProcess {
    const args = [
        "--enable-source-maps",
        "../../packages/matter-server/dist/esm/MatterServer.js",
        `--storage-path=${storagePath}`,
        `--log-level=${logLevel}`,
        `--port=${SERVER_PORT}`,
    ];
    if (enableTestNetDcl) {
        args.push("--enable-test-net-dcl");
    }
    if (logFilePath !== undefined) {
        args.push(`--log-file=${logFilePath}`);
    }
    const serverProcess = spawn("node", args, {
        cwd: process.cwd(),
        detached: true,
        stdio: ["pipe", "pipe", "pipe"],
    });

    serverProcess.stdout?.on("data", (data: Buffer) => {
        console.log("[server]", data.toString().trim());
    });
    serverProcess.stderr?.on("data", (data: Buffer) => {
        console.log("[server:err]", data.toString().trim());
    });

    return serverProcess;
}

/**
 * Starts the test device process with persistent stdout/stderr logging.
 */
export function startTestDevice(storagePath: string): ChildProcess {
    const proc = spawn(
        "npx",
        ["tsx", "test/fixtures/TestLightDevice.ts", `--storage-path=${storagePath}`, `--port=${DEVICE_PORT}`],
        {
            cwd: process.cwd(),
            detached: true, // own process group so all npx/tsx/node children are killed together
            stdio: ["pipe", "pipe", "pipe"],
        },
    );

    // Keep logging throughout the entire test run (not just during startup)
    proc.stdout?.on("data", (data: Buffer) => {
        console.log("[device]", data.toString().trim());
    });
    proc.stderr?.on("data", (data: Buffer) => {
        console.log("[device:err]", data.toString().trim());
    });

    return proc;
}

/**
 * Waits for a port to be ready by attempting WebSocket connections.
 */
export async function waitForPort(port: number, timeoutMs = 30_000): Promise<void> {
    const start = Date.now();
    while (Date.now() - start < timeoutMs) {
        try {
            const ws = new WebSocket(`ws://localhost:${port}/ws`);
            await new Promise<void>((resolve, reject) => {
                ws.on("open", () => {
                    ws.close();
                    resolve();
                });
                ws.on("error", reject);
            });
            return;
        } catch {
            await new Promise(r => setTimeout(r, 500));
        }
    }
    throw new Error(`Timeout waiting for port ${port}`);
}

/**
 * Waits for the device process to output readiness indicators.
 * Logging is handled by startTestDevice; this only resolves on the readiness signal.
 */
export function waitForDeviceReady(process: ChildProcess, timeoutMs = 30_000): Promise<void> {
    return new Promise((resolve, reject) => {
        const timeout = setTimeout(() => {
            process.stdout?.off("data", onData);
            reject(new Error("Timeout waiting for device to be ready"));
        }, timeoutMs);

        const onData = (data: Buffer) => {
            const output = data.toString();
            // Device is ready when it shows the pairing codes
            if (output.includes("Manual pairing code:") || output.includes("commissioned")) {
                clearTimeout(timeout);
                process.stdout?.off("data", onData);
                // Give it a moment to fully initialize network
                setTimeout(resolve, 2000);
            }
        };

        process.stdout?.on("data", onData);
    });
}

/**
 * Gracefully kills a process (and its process group if detached) and waits for exit.
 */
export async function killProcess(childProcess: ChildProcess | undefined, timeoutMs = 2_000): Promise<void> {
    if (!childProcess || childProcess.exitCode !== null) {
        if (childProcess) {
            childProcess.stdout?.removeAllListeners();
            childProcess.stderr?.removeAllListeners();
            childProcess.removeAllListeners();
        }
        return;
    }

    await new Promise<void>(resolve => {
        // Kill the whole process group so npx/tsx children don't outlive the parent.
        const sendSignal = (signal: "SIGINT" | "SIGKILL") => {
            if (childProcess.pid) {
                try {
                    nodeProcess.kill(-childProcess.pid, signal);
                } catch {
                    childProcess.kill(signal);
                }
            } else {
                childProcess.kill(signal);
            }
        };

        const cleanup = async () => {
            clearTimeout(timeout);
            childProcess.stdout?.removeAllListeners();
            childProcess.stderr?.removeAllListeners();
            childProcess.removeAllListeners();
            // Brief settle so the OS reclaims port bindings before the caller proceeds.
            await new Promise(r => setTimeout(r, 200));
            resolve();
        };

        const timeout = setTimeout(() => {
            if (childProcess.exitCode === null) {
                sendSignal("SIGKILL");
            }
            void cleanup();
        }, timeoutMs);

        childProcess.once("exit", () => void cleanup());
        sendSignal("SIGINT");
    });
}

File: matterjs-server-main/python_client/tests/helpers/process_helpers.py
"""Process management helpers for integration tests.

Ports the functionality of the JS ProcessHelpers.ts for use in Python integration tests.
Provides helpers to start/stop the Matter.js server and test device subprocesses.
"""

from __future__ import annotations

import asyncio
import contextlib
import os
from pathlib import Path
import shutil
import signal
import subprocess
import tempfile
import time

import aiohttp

REPO_ROOT = Path(__file__).parent.parent.parent.parent  # python_client/tests/helpers -> repo root
SERVER_PORT = 5580
SERVER_WS_URL = f"ws://localhost:{SERVER_PORT}/ws"
DEVICE_PORT = 5540
MANUAL_PAIRING_CODE = "34970112332"


def create_temp_storage_paths() -> tuple[str, str]:
    """Create temporary storage directories for server and device.

    Returns a tuple of (server_storage_path, device_storage_path).
    Both directories are created on disk immediately.
    """
    timestamp = int(time.time() * 1000)
    server_path = os.path.join(tempfile.gettempdir(), f"matter-test-server-{timestamp}")
    device_path = os.path.join(tempfile.gettempdir(), f"matter-test-device-{timestamp}")
    os.makedirs(server_path, exist_ok=True)
    os.makedirs(device_path, exist_ok=True)
    return server_path, device_path


def cleanup_temp_storage(server_path: str, device_path: str) -> None:
    """Remove temporary storage directories.

    Silently ignores errors if directories don't exist or can't be removed.
    """
    for path in (server_path, device_path):
        with contextlib.suppress(OSError):
            shutil.rmtree(path)


def start_server(storage_path: str, enable_test_net_dcl: bool = True) -> subprocess.Popen:
    """Start the Matter.js server process.

    Args:
        storage_path: Path to the temporary storage directory for server state.
        enable_test_net_dcl: Pass --enable-test-net-dcl so the server accepts test
            PAA certificates. Required when commissioning the TestLightDevice (which
            uses a test PAA). Set to False to test that commissioning fails without it.

    Returns:
        The subprocess.Popen handle for the running server.
    """
    server_script = REPO_ROOT / "packages" / "matter-server" / "dist" / "esm" / "MatterServer.js"
    args = [
        "node",
        "--enable-source-maps",
        str(server_script),
        f"--storage-path={storage_path}",
    ]
    if enable_test_net_dcl:
        args.append("--enable-test-net-dcl")
    return subprocess.Popen(
        args,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        cwd=str(REPO_ROOT),
    )


def start_test_device(storage_path: str) -> subprocess.Popen:
    """Start the test light device process.

    Args:
        storage_path: Path to the temporary storage directory for device state.

    Returns:
        The subprocess.Popen handle for the running test device.
    """
    device_script = (
        REPO_ROOT / "packages" / "matter-server" / "test" / "fixtures" / "TestLightDevice.ts"
    )
    return subprocess.Popen(
        [
            "npx",
            "tsx",
            str(device_script),
            f"--storage-path={storage_path}",
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        cwd=str(REPO_ROOT),
    )


async def wait_for_port(port: int, timeout: float = 30.0) -> None:
    """Wait for the server to be fully ready on the given port.

    Polls the HTTP /health endpoint until it returns 200 with a valid JSON body.
    This is stricter than a bare TCP/WebSocket connect because it confirms the
    server has finished initialising and all request handlers are wired up.

    Args:
        port: The port number to connect to.
        timeout: Maximum time to wait in seconds.

    Raises:
        TimeoutError: If the server doesn't become available within the timeout.
    """
    start = time.monotonic()
    url = f"http://localhost:{port}/health"
    while time.monotonic() - start < timeout:
        try:
            async with aiohttp.ClientSession() as session, asyncio.timeout(2):
                async with session.get(url) as resp:
                    if resp.status == 200:
                        body = await resp.json()
                        if "version" in body and "node_count" in body:
                            return
        except (TimeoutError, aiohttp.ClientError, OSError):
            pass
        await asyncio.sleep(0.5)
    raise TimeoutError(f"Timeout waiting for server /health on port {port}")


async def wait_for_device_ready(process: subprocess.Popen, timeout: float = 30.0) -> None:
    """Wait for the test device to be ready by monitoring its stdout.

    Watches for the "Manual pairing code:" or "commissioned" output lines
    that indicate the device is ready to accept connections.

    Args:
        process: The device subprocess to monitor.
        timeout: Maximum time to wait in seconds.

    Raises:
        RuntimeError: If the device process exits unexpectedly.
        TimeoutError: If the device doesn't become ready within the timeout.
    """
    assert process.stdout is not None
    loop = asyncio.get_running_loop()

    def _read_until_ready() -> None:
        """Blocking reader, run via run_in_executor to avoid stalling the event loop."""
        start = time.monotonic()
        while time.monotonic() - start < timeout:
            if process.stdout is None:
                raise RuntimeError("Process stdout is None")
            line = process.stdout.readline()
            if not line:
                if process.poll() is not None:
                    raise RuntimeError("Device process exited unexpectedly")
                time.sleep(0.1)
                continue
            if "Manual pairing code:" in line or "commissioned" in line.lower():
                return
        raise TimeoutError("Timeout waiting for device to be ready")

    await loop.run_in_executor(None, _read_until_ready)
    # Give the device's network stack time to fully initialize
    await asyncio.sleep(2)


def kill_process(process: subprocess.Popen | None, timeout: float = 5.0) -> None:
    """Gracefully kill a process, falling back to SIGKILL if needed.

    Sends SIGINT first to allow graceful shutdown, then SIGKILL if the
    process doesn't exit within the timeout.

    Args:
        process: The process to kill, or None (no-op).
        timeout: Time to wait for graceful shutdown before forcing kill.
    """
    if process is None or process.poll() is not None:
        return
    process.send_signal(signal.SIGINT)
    try:
        process.wait(timeout=timeout)
    except subprocess.TimeoutExpired:
        process.kill()
        with contextlib.suppress(subprocess.TimeoutExpired):
            process.wait(timeout=2)
